### PR TITLE
Navigable Spectrum Display

### PR DESCRIPTION
**Pull request by amuskens**
Nearly finished. Only a few fixes left.

### CODE REVIEW

File: src/odemis/gui/comp/legend.py [hunk 1]
```diff
@@ -262,6 +262,8 @@ def __init__(self, parent, wid=wx.ID_ANY, pos=(0, 0), size=wx.DefaultSize,
         self._max_tick_width = 32  # Largest pixel width of any label in use
         self._tick_spacing = 120 if orientation == wx.HORIZONTAL else 80
         self._unit = None
+        self._lo_ellipsis = False
```
Comment (pieleric) [inline src/odemis/gui/comp/legend.py]: Add a comment to indicate what they mean. You could also add a docstring to the class and explain how the two attributes .lo_ellipsis and .hi_ellipsis are to be used.

File: src/odemis/gui/comp/legend.py [hunk 2]
```diff
@@ -281,12 +283,32 @@ def __init__(self, parent, wid=wx.ID_ANY, pos=(0, 0), size=wx.DefaultSize,
...
             self.Refresh()
 
+    @lo_ellipsis.setter
+    def lo_ellipsis(self, val):
+        if self._lo_ellipsis != val:
+            self._lo_ellipsis = val
+            self.Refresh()
+
+    @hi_ellipsis.setter
+    def hi_ellipsis(self, val):
```
Comment (pieleric) [inline src/odemis/gui/comp/legend.py]: While at it, also check that `val` is a boolean.

File: src/odemis/gui/comp/legend.py [hunk 3]
```diff
@@ -347,6 +369,13 @@ def on_paint(self, _):
 
         for i, (pos, val) in enumerate(self._tick_list):
             label = units.readable_str(val, self.unit, sig)
+
+            if i == 0 and self._lo_ellipsis:
+                label = "..." + label
```
Comment (pieleric) [inline src/odemis/gui/comp/legend.py]: There is a special ellipsis character, which might display better with some fonts. So you might want to use it: `u"…"`.

File: src/odemis/gui/comp/miccanvas.py [hunk 4]
```diff
@@ -1165,6 +1166,110 @@ def update_drawing(self):
...
+        super(NavigableBarPlotCanvas, self).__init__(*args, **kwargs)
+        self.abilities |= {CAN_DRAG}
+        
+        # When resizing, margin to put around the current size
+        # TODO: Maybe make the margin related to the canvas size?
+        self.default_margin = 512
+        self.margins = (self.default_margin, self.default_margin)
+
+        # the position the view is asking to the next buffer recomputation
+        # in buffer-coordinates: = 1px at scale = 1
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: What is this comment about? Probably not `data_buffer`. Is that `window_x/y`? => Move to the right place, or make it more explicit in the comments.

File: src/odemis/gui/comp/miccanvas.py [hunk 5]
```diff
@@ -1165,6 +1166,110 @@ def update_drawing(self):
...
+        self._data_buffer = None
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: Preferably, for each attribute initialised with `None`, add a comment indicating the data type and unit.

File: src/odemis/gui/comp/miccanvas.py [hunk 6]
```diff
@@ -1165,6 +1166,110 @@ def update_drawing(self):
...
+        self.window_x = None
+        self.window_y = None
+        self.display_xrange = None
+        self.display_yrange = None
+        self.data_xrange = None
+        self.data_yrange = None
+
+    def set_1d_data(self, xs, ys, unit_x=None, unit_y=None, range_x=None, range_y=None):
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: Is calling `set_data()` still possible? (I guess not really). If not, please mention it in the docstring.

File: src/odemis/gui/comp/miccanvas.py [hunk 7]
```diff
@@ -1165,6 +1166,110 @@ def update_drawing(self):
...
+        self.reset_ranges()
+
+    def reset_ranges(self):
+        if self._data_buffer is None:
+            return
+
+        self.set_ranges(self.data_xrange, self.data_yrange)
+
+    def set_ranges(self, x_range, y_range):
+
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: Needs a docstring.

File: src/odemis/gui/comp/miccanvas.py [hunk 8]
```diff
@@ -1165,6 +1166,110 @@ def update_drawing(self):
...
+        if self._data_buffer is None:
+            return
+
+        (xst , yst) = self._data_buffer
+
+        # Deal with x
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: Please explain more the goal of this code. It's to cache just the data needed to be displayed in the selected range. Something like this, right?

File: src/odemis/gui/comp/miccanvas.py [hunk 9]
```diff
@@ -1165,6 +1166,110 @@ def update_drawing(self):
...
+        # Deal with Y
+        if y_range is not None:
+            lo, hi = y_range
+            ys = numpy.clip(ys, lo, hi)
+
+        self.display_xrange = x_range
+        self.display_yrange = y_range
+
+        # Resample data based on window size
+        if len(xs) > self.ClientSize.x * 10:
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: If you're using the ClientSize, it means that probably this function (or a similar one) should called whenever the ClientSize changes. However, I don't see such code... so probably the display will go wrong on the window resize.

File: src/odemis/gui/comp/viewport.py [hunk 10]
```diff
@@ -1065,7 +1066,175 @@ def _on_peak_method(self, state):
...
+        # dragging!
+        self._hdragging = False
+        self._hdrag_scale_factor = -0.1
+
+        # The amount of pixels shifted in the current drag event
+        self.drag_shift = (0, 0)  # px, px
+        #  initial position of mouse when started dragging
+        self.drag_init_pos = (0, 0)  # px, px
+        self._last_pos = (0, 0)
+        self._last_time = 0
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: Add comments about the data type and unit of these two attributes. Also, explain what `last_time` is about. (If I understand correctly, it's the last time the mouse move during a drag, to infer the speed, to detect when to reset... something?)

File: src/odemis/gui/comp/viewport.py [hunk 11]
```diff
@@ -1065,7 +1066,175 @@ def _on_peak_method(self, state):
...
+
+        # Track if the legend was dragged. It should be reset to False at the
+        # end of button up handlers, giving overlay the chance to check if a
+        # drag occurred.
+        self.was_dragged = False
+
+        self.canvas.Bind(wx.EVT_MOUSEWHEEL, self.on_hlegend_scroll)
+        self.canvas.Bind(wx.EVT_MIDDLE_DOWN, self.on_hlegend_drag_start)
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: As you also bind the canvas, I'd suggest to rename all these functions to `on_hor_*()` and `on_ver_*()`. If you leave them named `...legend...`, at some point, we'll forget that they can be called when the canvas is receiving event, and the wrong assumptions will be done.

File: src/odemis/gui/comp/viewport.py [hunk 12]
```diff
@@ -1065,7 +1066,175 @@ def _on_peak_method(self, state):
...
+        """
+        if self._hdragging:
+
+            v_pos = evt.Position
+            now = time.time()
+            speed = ((v_pos[0] - self._last_pos[0]) / (now - self._last_time),
+                     (v_pos[1] - self._last_pos[1]) / (now - self._last_time))
+
+            # if the mouse stopped, reset the init position
+            if abs(speed[0]) < 0.05:
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: I don't understand the goal of resetting the drag_init_pos if the user is not fast enough. In general I think it's dangerous to rely  on speed, because some users might be slower that others.

File: src/odemis/gui/comp/viewport.py [hunk 13]
```diff
@@ -1065,7 +1066,175 @@ def _on_peak_method(self, state):
...
+        # end of button up handlers, giving overlay the chance to check if a
+        # drag occurred.
+        self.was_dragged = False
+
+        self.canvas.Bind(wx.EVT_MOUSEWHEEL, self.on_hlegend_scroll)
+        self.canvas.Bind(wx.EVT_MIDDLE_DOWN, self.on_hlegend_drag_start)
+        self.canvas.Bind(wx.EVT_MIDDLE_UP, self.on_hlegend_drag_end)
+        self.canvas.Bind(wx.EVT_MOTION, self.on_hlegend_motion)
+        self.bottom_legend.Bind(wx.EVT_MOUSEWHEEL, self.on_hlegend_scroll)
+        self.left_legend.Bind(wx.EVT_MOUSEWHEEL, self.on_vlegend_scroll)
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: Also need to handle drag on the left legend, or is it already handle somewhere else?

File: src/odemis/gui/util/img.py [hunk 14]
```diff
@@ -929,7 +929,8 @@ def calculate_ticks(value_range, client_size, orientation, tick_spacing):
                                orientation)
         # Round "almost 0" (due to floating point errors) to 0
         if abs(tick_value) < epsilon:
-            tick_value = 0
+            # tick_value = 0
```
Comment (pieleric) [inline src/odemis/gui/util/img.py]: Why is that needed? It should eventually go back. Maybe you need to tweak the computation of `epsilon`?

**Comment by amuskens** [comment]
Still one glitch: the VA does not seem to link properly to the axis lock toggle. Will see if I can fix this.

**Comment by amuskens** [comment]
Note: Still some strange truncation in the vertical plots that are exported. Will investigate further.

**Comment by amuskens** [comment]
I believe it is ready for a review.

File: src/odemis/gui/comp/miccanvas.py [hunk 15]
```diff
@@ -1165,6 +1172,182 @@ def update_drawing(self):
...
+
+        if lox != hix:
+            xs = xst[lox:hix]
+            ys = yst[lox:hix]
+        else:
+            xs = [(hi - lo) / 2 + lo]
+            ys = [yst[lox]]
+
+        # Add a few points onto the beginning and end of the array
+        # to prevent gaps in teh data from appearing
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: Typo: teh -> the.

File: src/odemis/gui/comp/miccanvas.py [hunk 16]
```diff
@@ -1165,6 +1172,182 @@ def update_drawing(self):
...
+        lo, hi = self.display_xrange
+
+        # w = numpy.where((xst > lo) * (xst < hi))
+        lox = numpy.searchsorted(xst, lo, side="left")
+        hix = numpy.searchsorted(xst, hi, side="right")
+
+        if lox != hix:
+            xs = xst[lox:hix]
+            ys = yst[lox:hix]
+        else:
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: Please describe when this happens (zoomed so much that there is only one point?), and what is done is such case (show a single bar in the middle???)

File: src/odemis/gui/comp/miccanvas.py [hunk 17]
```diff
@@ -1165,6 +1172,182 @@ def update_drawing(self):
...
+            xs = [(hi - lo) / 2 + lo]
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: That also can be written `(hi + lo) / 2`, right?

File: src/odemis/gui/comp/miccanvas.py (hunk 8 above)
Comment (pieleric) [reply src/odemis/gui/comp/miccanvas.py]: You still need to add a comment here explaining what's happening here. "Deal with x" is not enough!

File: src/odemis/gui/comp/miccanvas.py [hunk 18]
```diff
@@ -1165,6 +1172,182 @@ def update_drawing(self):
...
+        self.display_yrange = y_range
+
+        self.refresh_plot()
+
+        self.Refresh()
+
+    def on_size(self, evt):
+        # Reset the data display to ensure the data resampling is done
+        self.refresh_plot()
+        BarPlotCanvas.on_size(self, evt)
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: Use `super()` please.

File: src/odemis/gui/comp/miccanvas.py [hunk 19]
```diff
@@ -1165,6 +1172,182 @@ def update_drawing(self):
...
+
+        if y_range is None:
+            y_range = self.data_yrange
+
+        self.display_xrange = x_range
+        self.display_yrange = y_range
+
+        self.refresh_plot()
+
+        self.Refresh()
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: No need for `Refresh()`, as that's already done by `refresh_plot()`.

File: src/odemis/gui/comp/miccanvas.py [hunk 20]
```diff
@@ -1165,6 +1172,182 @@ def update_drawing(self):
...
+            xs = numpy.insert(xs, 0, xst[lox - 1])
+            ys = numpy.insert(ys, 0, yst[lox - 1])
+
+        if hix < len(xst) and xs[-1] != hi:
+            xs = numpy.append(xs, xst[hix])
+            ys = numpy.append(ys, yst[hix])
+
+        # Re-sample data based on window size
+        if len(xs) > self.ClientSize.x * 10:
+            rs_factor = len(xs) // self.ClientSize.x // 7
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: I'd suggest to move this entire block into a separate function, so that it just looks like:
`xs, ys = self._resample_plot(xs, ys, self.ClientSize.x * 7)`
(The 3rd argument is the new approximate length).

File: src/odemis/gui/comp/miccanvas.py [hunk 21]
```diff
@@ -1165,6 +1172,227 @@ def update_drawing(self):
...
+
+    def _resample_plot(self, xs, ys, threshold):
+        """
+        Re-sample the data if there are more points than the threshold
+        xs, ys: two 1-D arrays with data
+        threshold: the minimum number of points for a re-sampling to occur
+        """
+        # Re-sample data based on window size
+        if len(xs) > threshold:
+            rs_factor = len(xs) // self.ClientSize.x // 6
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: I'd rather have the function just contain the content of this block (and pass `self.ClientSize.x * 6` as argument). The main advantage is that it'll not relay on  `self.ClientSize.x`, so it'll be generic, can be more easily reused. The other advantage is that it'd ensure that it's not possible to write `len(xs) // self.ClientSize.x // 6`, which is unreadable ;-) Instead, it'd be be `rs_factor = len(xs) // new_len`.

File: src/odemis/gui/comp/legend.py [hunk 22]
```diff
@@ -307,6 +385,20 @@ def on_size(self, _=None):
         if self._value_range:
             self.Refresh()
 
+        if self._orientation == wx.HORIZONTAL:
+            self.lockBtn.SetPosition((self.ClientSize.x - 24, 0))
+        else:
+            self.lockBtn.SetPosition((self.ClientSize.x - self.ClientSize.x / 2 - 12, 0))
+
+    def on_mouse_enter(self, _=None):
```
Comment (pieleric) [inline src/odemis/gui/comp/legend.py]: Is it really needed to update the cursor every time the mouse enters/leaves? IIRC, with wxPython, SetCursor is local to the widget any way. So you could just change the cursor at init, and be done.

File: src/odemis/gui/comp/miccanvas.py [hunk 23]
```diff
@@ -36,10 +36,12 @@
     call_in_wx_main
 from odemis.gui.util.img import format_rgba_darray
 from odemis.util import units
-import scipy.ndimage
+import scipy.ndimage, scipy.signal
```
Comment (pieleric) [inline src/odemis/gui/comp/miccanvas.py]: You don't need this import any more, do you?

File: src/odemis/gui/comp/viewport.py [hunk 24]
```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
...
+
+    Note: clamping of the ranges is handle by the mouse events, because
+    the clamping behavior must function differently for pans, drags, and scrolling
+
+    """
+    __metaclass__ = no_conflict.classmaker(right_metas=(ABCMeta,))
+
+    # Default class
+    canvas_class = miccanvas.NavigableBarPlotCanvas
+    bottom_legend_class = AxisLegend
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: No need to override these two legend classes, as they are the same as in PlotViewport.

File: src/odemis/gui/comp/viewport.py [hunk 25]
```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
...
+            hi = self.canvas.data_yrange[1] + self.canvas.data_yrange[1] * self._vmargin
+
+        self.vrange.value = (lo, hi)
+
+    def on_drag_start(self, evt):
+        """ Start a dragging procedure """
+        if not self.canvas.has_data():
+            return
+
+        # Fixme: only go to drag mode if the mouse moves before a mouse up?
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: No, I don't think this is needed. Or what would be the advantage? => drop the comment?

File: src/odemis/gui/comp/viewport.py [hunk 26]
```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
...
+    def on_drag_end(self, evt):
+        """ End the dragging procedure """
+        if not self.canvas.has_data():
+            return
+
+        # If the canvas was being dragged
+        if self._dragging:
+            self._dragging = False
+
+        self.was_dragged = False
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: No one cares (apparently) => drop.

File: src/odemis/gui/comp/viewport.py [hunk 27]
```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
...
+    def on_canvas_motion(self, evt):
+        """ Process mouse motion
+
+        Set the drag shift and refresh the image if dragging is enabled and the left mouse button is
+        down.
+
+        """
+        if self._dragging:
+
+            v_pos = (self.canvas.pos_x_to_val_x(evt.Position[0]),
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: Isn't it possible to use `v_pos = self.canvas.pos_to_val(evt.Position)`?

File: src/odemis/gui/comp/viewport.py [hunk 28]
```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
...
+
+        self.vrange = model.TupleVA(setter=self.update_vrange)
+        self.vrange_lock = model.BooleanVA(False)
+        self.left_legend.lock_va = self.vrange_lock
+
+        # Indicate a left mouse button drag in the canvas
+        # Note: *only* use it indicate that the *canvas* is performing an operation related to
+        # dragging!
+        self._dragging = False
+        self._hdrag_scale_factor = -0.1
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: What are these scale factors about? I somehow imagine that the "-" is to make sure that it goes in the same direction as the mouse, while dragging with the legend causes the opposite direction move... But they all have this one, so I'm confused. And why dividing by 10? Is that to do a small move? If so, why not making sure that it respects the position of the mouse?

File: src/odemis/gui/test/comp_viewport_test.py [hunk 29]
```diff
@@ -150,6 +150,9 @@ def test_plot_viewport(self):
         for mode in MODES:
             vwp.canvas.set_plot_mode(mode)
 
+            """
```
Comment (pieleric) [inline src/odemis/gui/test/comp_viewport_test.py]: Are they really no way to provide bad ranges? Maybe if rng[1] < rng[0] ?

File: src/odemis/gui/comp/miccanvas.py (hunk 21 above)
Comment (pieleric) [reply src/odemis/gui/comp/miccanvas.py]: Now I see that `threshold` is actually computed as `self.ClientSize.x * 10`. So you can just compute the rs_factor as ` len(xs) // threshold`, or  ` len(xs) // (threshold / 2)`

File: src/odemis/gui/util/img.py [hunk 30]
```diff
@@ -1292,6 +1340,38 @@ def chronogram_to_export_data(proj, raw):
         return spec
     else:
         # Draw spectrum bar plot
+        if vp is not None:
```
Comment (pieleric) [inline src/odemis/gui/util/img.py]: Please factorize into a common function for both the spectrum data and time data.

File: src/odemis/gui/util/img.py [hunk 31]
```diff
@@ -1159,12 +1160,13 @@ def bar_plot(ctx, data, data_width, range_x, data_height, range_y, client_size,
...
 
 
-def spectrum_to_export_data(proj, raw):
+def spectrum_to_export_data(proj, raw, vp=None):
     """
     Creates either raw or WYSIWYG representation for the spectrum data plot
 
     proj (DataProjection): DataProjection of the spectrum to export
     raw (boolean): if True returns raw representation
+    vp (Viewport): Viewport object
```
Comment (pieleric) [inline src/odemis/gui/util/img.py]: Note: ideally, you wouldn't pass a viewport, but a `gui.model.StreamView`. That (special) view would have the corresponding VAs `hrange` and `vrange`. The NavigablePlotViewport would just read/write them, whenever the user moves or zooms around. Anyway, for now, let's keep it like this.

File: src/odemis/gui/comp/viewport.py [hunk 32]
```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
...
+    def on_vlegend_motion(self, evt):
+        """ Process mouse motion
+
+        Set the drag shift and refresh the image if dragging is enabled and the left mouse button is
+        down.
+
+        """
+        if self._dragging:
+
+            v_pos = (self.canvas.pos_x_to_val_x(evt.Position[0]),
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: You don't really need `v_pos[0]`, do you? If so, you can skip its computation. Same thing for `on_hlegend_motion()`

File: src/odemis/gui/comp/viewport.py [hunk 33]
```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
...
+
+        logging.debug("Drag started at %s", self.drag_init_pos)
+
+    def on_drag_end(self, evt):
+        """ End the dragging procedure """
+        if not self.canvas.has_data():
+            return
+
+        # If the canvas was being dragged
+        if self._dragging:
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: You can just always do `self._dragging = False`.

File: src/odemis/gui/comp/viewport.py [hunk 34]
```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
...
+
+        pos = evt.Position
+        self.drag_init_pos = (self.canvas.pos_x_to_val_x(pos[0]),
+                              self.canvas.pos_y_to_val_y(pos[1]))
+
+        logging.debug("Drag started at %s", self.drag_init_pos)
+
+    def on_drag_end(self, evt):
+        """ End the dragging procedure """
+        if not self.canvas.has_data():
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: No need to check.

File: src/odemis/gui/comp/viewport.py [hunk 35]
```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
...
+        if rot < 0:
+            new_span = span * (1 + scale)
+        else:
+            new_span = span * (1 - scale)
+
+        lo = zoom_centre - prop * new_span
+        hi = lo + new_span
+
+        # Clamp the ranges - this functions differently for scrolls than pans
+        if lo < self.canvas.data_yrange[0]:
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: For such simple clip/clamp, you can just write like:
```
lo = max(self.canvas.data_yrange[0], lo)
hi = min(hi, self.canvas.data_yrange[1] + self.canvas.data_yrange[1] * self._vmargin)
```

BTW, be aware that this margin computation works only because yrange[0] = 0 (and yrange[1] is always > 0). I think the complete way to write it is `self.canvas.data_yrange[1] + (self.canvas.data_yrange[1] - self.canvas.data_yrange[0]) * self._vmargin`.

**Comment by pieleric** [comment]
BTW, for the final merge, it'd be nice if you could have just two commits: one adding support to the GUI, and one extending the export.

File: src/odemis/gui/comp/viewport.py (hunk 28 above)
Comment (amuskens) [reply src/odemis/gui/comp/viewport.py]: Hi. I found that somehow the canvas dragged too fast without dividing by 10, and it was a bit awkward for the user to control.

File: src/odemis/gui/comp/viewport.py (hunk 27 above)
Comment (amuskens) [reply src/odemis/gui/comp/viewport.py]: Apparently it doesn't have it - which is strange. I left it the way it was.

**Comment by amuskens** [comment]
Made changes. Let me know when it is ready to re-base.

**Comment by pieleric** [comment]
It now looks good. Could you try to rebase the changes into 2 commits (viewport vs export)?

**Comment by amuskens** [comment]
Rebase complete.
